pkg/restapi/vc/operation/operations.go

```diff
@@ -322,41 +361,90 @@ func (o *Operation) storeVCHandler(rw http.ResponseWriter, req *http.Request) {
 		return
 	}
 
-	doc := operation.StructuredDocument{}
-	doc.Content = make(map[string]interface{})
-	doc.Content["message"] = data.Credential
+	o.storeVC(data, vc, rw)
+}
 
-	// We need an EDV-compliant ID. The UUID that's in the VC is a good fit, since it's 128 bits long already.
-	// Let's convert it to base58 and use it as the StructuredDocument ID.
-	edvDocID, err := convertVCIDToEDVCompatibleFormat(vc.ID)
+func (o *Operation) storeVC(data *StoreVCRequest, vc *verifiable.Credential, rw http.ResponseWriter) {
+	doc, err := o.buildStructuredDoc(data, vc)
 	if err != nil {
 		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
 
 		return
 	}
 
-	doc.ID = edvDocID
+	marshalledStructuredDoc, err := json.Marshal(doc)
+	if err != nil {
+		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
+
+		return
+	}
+
+	// We have no recipients, so we pass in the sender key as the recipient key as well
+	encryptedStructuredDoc, err := o.packer.Pack(marshalledStructuredDoc,
+		base58.Decode(o.senderKey), [][]byte{base58.Decode(o.senderKey)})
+	if err != nil {
+		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
+
+		return
+	}
+
+	encryptedDocument := operation.EncryptedDocument{
+		ID:       doc.ID,
+		Sequence: 0,
+		JWE:      encryptedStructuredDoc,
+	}
+
+	_, err = o.client.CreateDocument(data.Profile, &encryptedDocument)
+	if err != nil {
+		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
+
+		return
+	}
 
-	_, err = o.client.CreateDocument(data.Profile, &doc)
+	err = o.idMapping.Put(vc.ID, []byte(doc.ID))
 	if err != nil {
 		o.writeErrorResponse(rw, http.StatusBadRequest, err.Error())
 
 		return
 	}
 }
 
-// This function expects the vcID that's passed in to be a URL with a "/" and then a UUID concatenated to the end.
-func convertVCIDToEDVCompatibleFormat(vcID string) (string, error) {
-	vcIDParts := strings.Split(vcID, "/")
+func (o *Operation) buildStructuredDoc(data *StoreVCRequest,
+	vc *verifiable.Credential) (*operation.StructuredDocument, error) {
+	var edvDocID string
 
-	uuidFromVCID := vcIDParts[len(vcIDParts)-1]
+	idFromMapping, err := o.idMapping.Get(vc.ID)
+	if err != nil {
+		edvDocID, err = generateEDVCompatibleID()
```

Comment:
we lost the original `err` here, are you sure you don't want to wrap/embed the 2 errors.. or at least log a warning with the original one?

since the original err is from a `Get()` function, I would assume you want to call `generateEDVCompatibleID` only if `Get()` returned a `notFound` error and other errors are valid error type that should return automatically .. 

this logic needs to be revisited.

Reply:
Ah, when I first did this I had falsely assumed that the Aries storage didn't have a generic "not found" error, but now that I take a closer look it does indeed! I've updated the code to check for the specific condition now.

Reply:
Just as a heads up: this now uses the Value Not Found error from edge-core. Same idea. Works the same way.